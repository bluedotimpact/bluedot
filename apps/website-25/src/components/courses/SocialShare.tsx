import React from 'react';
import { FaFacebook, FaLinkedin, FaTwitter } from 'react-icons/fa6';

type SocialShareProps = {
  courseUrl: string;
  referralCode?: string;
  text?: string;
};

const SocialShare: React.FC<SocialShareProps> = ({ courseUrl, referralCode, text }) => {
  return (
    <div className="social-share flex flex-row gap-4">
      <a
        className="social-share__link size-6"
        href={`https://www.linkedin.com/shareArticle?mini=true&url=${courseUrl}?r=${referralCode}&utm_source=referral&utm_campaign=linkedin&text=${text}`}
        target="_blank"
        rel="noopener noreferrer"
      >
        <FaLinkedin className="social-share__link-icon size-6" />
      </a>
      <a
        className="social-share__link size-6"
        href={`https://twitter.com/intent/tweet?text=${text}&url=${courseUrl}?r=${referralCode}`}
        target="_blank"
        rel="noopener noreferrer"
      >
        <FaTwitter className="social-share__link-icon size-6" />
      </a>
      <a
        className="social-share__link size-6"
        href={`https://www.facebook.com/sharer/sharer.php?u=${courseUrl}?r=${referralCode}&utm_source=referral&utm_campaign=fb&quote=${text}`}
        target="_blank"
        rel="noopener noreferrer"
      >
        <FaFacebook className="social-share__link-icon size-6" />
      </a>
    </div>
  );
};

export default SocialShare;
