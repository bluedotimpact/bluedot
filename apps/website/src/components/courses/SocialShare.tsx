import { addQueryParam, ClickTarget, useAuthStore } from '@bluedot/ui';
import React from 'react';
import { FaFacebook, FaLinkedin, FaTwitter } from 'react-icons/fa6';
import useAxios from 'axios-hooks';

type SocialShareProps = {
  coursePath: string;
  text?: string;
};

type GetReferralResponse = {
  type: 'success';
  referralId: string;
};

const SocialShare: React.FC<SocialShareProps> = ({ coursePath, text }) => {
  const auth = useAuthStore((s) => s.auth);

  const [{ data }] = useAxios<GetReferralResponse>({
    method: 'get',
    url: '/api/referrals',
    headers: {
      Authorization: `Bearer ${auth?.token}`,
    },
  });

  const constructFullCourseUrl = (campaign: string) => {
    const baseUrl = window.location.origin;
    const url = `${baseUrl}${coursePath}?utm_source=referral&utm_campaign=${campaign}`;
    return data?.referralId ? addQueryParam(url, 'r', data.referralId) : url;
  };

  return (
    <div className="social-share flex flex-row gap-4">
      <ClickTarget
        className="social-share__link size-6"
        url={`https://www.linkedin.com/shareArticle?mini=true&url=${encodeURIComponent(constructFullCourseUrl('linkedin'))}${text ? `&text=${text}` : ''}`}
        target="_blank"
      >
        <FaLinkedin className="social-share__link-icon size-6" />
      </ClickTarget>
      <ClickTarget
        className="social-share__link size-6"
        url={`https://twitter.com/intent/tweet?url=${encodeURIComponent(constructFullCourseUrl('twitter'))}${text ? `&text=${text}` : ''}`}
        target="_blank"
      >
        <FaTwitter className="social-share__link-icon size-6" />
      </ClickTarget>
      <ClickTarget
        className="social-share__link size-6"
        url="https://www.facebook.com/sharer/sharer.php"
        target="_blank"
      >
        <FaFacebook className="social-share__link-icon size-6" />
      </ClickTarget>
    </div>
  );
};

export default SocialShare;
