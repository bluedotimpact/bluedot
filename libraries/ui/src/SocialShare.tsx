import React from 'react';
import {
  FaEnvelope,
  FaGithub,
  FaXTwitter,
  FaLinkedin,
  FaFacebook,
} from 'react-icons/fa6';
import { ClickTarget } from './ClickTarget';

type SocialShareProps = {
  /** The path to append to origin for sharing (e.g. /courses/future-of-ai) */
  path?: string;
  /** The text to share */
  text?: string;
  /** Whether this is for contact links or sharing */
  variant?: 'default' | 'contact';
  /** Optional referral ID to append to share URLs */
  referralId?: string;
  /** Custom links for contact mode */
  emailLink?: string;
  githubOrgLink?: string;
  twitterLink?: string;
  linkedinLink?: string;
  facebookLink?: string;
};

export const SocialShare: React.FC<SocialShareProps> = ({
  path,
  text,
  variant = 'default',
  referralId,
  emailLink,
  githubOrgLink,
  twitterLink,
  linkedinLink,
  facebookLink,
}) => {
  const constructShareUrl = (campaign?: string) => {
    if (variant === 'contact') return '';

    const baseUrl = typeof window !== 'undefined' ? window.location.origin : '';
    const fullPath = path || (typeof window !== 'undefined' ? window.location.pathname : '');
    let url = `${baseUrl}${fullPath}`;

    if (campaign) {
      url = `${url}?utm_source=referral&utm_campaign=${campaign}`;
      if (referralId) {
        url = `${url}&r=${referralId}`;
      }
    }

    return url;
  };

  const getShareUrl = (platform: string): string => {
    if (variant === 'contact') {
      switch (platform) {
        case 'email': return emailLink || '';
        case 'github': return githubOrgLink || '';
        case 'twitter': return twitterLink || '';
        case 'linkedin': return linkedinLink || '';
        case 'facebook': return facebookLink || '';
        default: return '';
      }
    }

    const shareUrl = constructShareUrl(platform);
    switch (platform) {
      case 'facebook':
        return `https://www.facebook.com/sharer/sharer.php?u=${encodeURIComponent(shareUrl)}`;
      case 'twitter':
        return `https://twitter.com/intent/tweet?url=${encodeURIComponent(shareUrl)}${text ? `&text=${encodeURIComponent(text)}` : ''}`;
      case 'linkedin':
        return `https://www.linkedin.com/shareArticle?mini=true&url=${encodeURIComponent(shareUrl)}${text ? `&text=${encodeURIComponent(text)}` : ''}`;
      default:
        return '';
    }
  };

  const baseClasses = 'flex items-center justify-center size-12';
  const variantClasses = variant === 'contact'
    ? 'rounded-full bg-gray-900 text-white hover:bg-gray-700 transition-colors'
    : 'text-gray-700 hover:text-gray-900 transition-colors';

  return (
    <div className="flex justify-center gap-4">
      {variant === 'contact' && emailLink && (
        <ClickTarget
          url={getShareUrl('email')}
          target="_blank"
          className={`${baseClasses} ${variantClasses}`}
          aria-label="Email"
        >
          <FaEnvelope size={20} />
        </ClickTarget>
      )}
      {variant === 'contact' && githubOrgLink && (
        <ClickTarget
          url={getShareUrl('github')}
          target="_blank"
          className={`${baseClasses} ${variantClasses}`}
          aria-label="GitHub"
        >
          <FaGithub size={20} />
        </ClickTarget>
      )}
      {(variant === 'default' || twitterLink) && (
        <ClickTarget
          url={getShareUrl('twitter')}
          target="_blank"
          className={`${baseClasses} ${variantClasses}`}
          aria-label="X (Twitter)"
        >
          <FaXTwitter size={20} />
        </ClickTarget>
      )}
      {(variant === 'default' || linkedinLink) && (
        <ClickTarget
          url={getShareUrl('linkedin')}
          target="_blank"
          className={`${baseClasses} ${variantClasses}`}
          aria-label="LinkedIn"
        >
          <FaLinkedin size={20} />
        </ClickTarget>
      )}
      {(variant === 'default' || facebookLink) && (
        <ClickTarget
          url={getShareUrl('facebook')}
          target="_blank"
          className={`${baseClasses} ${variantClasses}`}
          aria-label="Facebook"
        >
          <FaFacebook size={20} />
        </ClickTarget>
      )}
    </div>
  );
};
