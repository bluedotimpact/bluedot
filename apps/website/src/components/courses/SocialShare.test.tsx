import { render } from '@testing-library/react';
import {
  beforeEach,
  describe,
  expect,
  test,
  vi,
} from 'vitest';
import SocialShare from './SocialShare';

const mockReferralId = 'test-referral-123';

// Mock implementations
const mockAxiosHooks = vi.fn();
const mockUseAuthStore = vi.fn();

// Mock modules
vi.mock('axios-hooks', () => ({
  default: () => mockAxiosHooks(),
}));

vi.mock('@bluedot/ui', async () => {
  const actual = await vi.importActual('@bluedot/ui');
  return {
    ...actual,
    useAuthStore: () => mockUseAuthStore(),
  };
});

describe('SocialShare', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('when user is logged in', () => {
    beforeEach(() => {
      mockUseAuthStore.mockReturnValue({ token: 'mockToken', expiresAt: Date.now() + 10000 });
      mockAxiosHooks.mockReturnValue([{
        data: {
          referralId: mockReferralId,
        },
      }]);
    });

    test('renders default as expected', () => {
      const { container } = render(
        <SocialShare
          coursePath="/courses/future-of-ai"
          text="I've just completed a free, 2-hour course on the future of AI and its impacts on society. Here are my takeaways:"
        />,
      );
      expect(container).toMatchSnapshot();
    });

    test('renders without optional args', () => {
      const { container } = render(
        <SocialShare
          coursePath="/future-of-ai"
        />,
      );
      const socialShareLinks = container.querySelectorAll('.social-share__link');
      socialShareLinks.forEach((link) => {
        expect(link.getAttribute('href')).toMatchSnapshot();
      });
    });

    test('includes referralCode in share URLs', () => {
      const coursePath = '/courses/testing-referrals';
      const { container } = render(
        <SocialShare
          coursePath={coursePath}
        />,
      );

      const expectedLinks = [
        {
          baseUrl: 'linkedin.com',
          assert: (link: Element | null) => {
            const href = link?.getAttribute('href');
            expect(href).not.toBeNull();

            const urlParams = new URLSearchParams(href!.split('?')[1]);
            const encodedTargetUrl = urlParams.get('url');
            expect(encodedTargetUrl).not.toBeNull();
            if (encodedTargetUrl) {
              const targetUrl = new URL(decodeURIComponent(encodedTargetUrl));
              expect(targetUrl.searchParams.get('r')).toBe(mockReferralId);
              expect(targetUrl.searchParams.get('utm_source')).toBe('referral');
              expect(targetUrl.searchParams.get('utm_campaign')).toBe('linkedin');
            }
          },
        },
        {
          baseUrl: 'twitter.com',
          assert: (link: Element | null) => {
            const href = link?.getAttribute('href');
            expect(href).not.toBeNull();

            const urlParams = new URLSearchParams(href!.split('?')[1]);
            const encodedTargetUrl = urlParams.get('url');
            expect(encodedTargetUrl).not.toBeNull();
            if (encodedTargetUrl) {
              const targetUrl = new URL(decodeURIComponent(encodedTargetUrl));
              expect(targetUrl.searchParams.get('r')).toBe(mockReferralId);
              expect(targetUrl.searchParams.get('utm_source')).toBe('referral');
              expect(targetUrl.searchParams.get('utm_campaign')).toBe('twitter');
            }
          },
        },
        {
          baseUrl: 'facebook.com',
          assert: (link: Element | null) => {
            const href = link?.getAttribute('href');
            expect(href).not.toBeNull();
          },
        },
      ];

      expectedLinks.forEach(({ baseUrl, assert }) => {
        const link = container.querySelector(`a[href*="${baseUrl}"]`);
        assert(link);
      });
    });
  });

  describe('when user is logged out', () => {
    beforeEach(() => {
      mockUseAuthStore.mockReturnValue({ token: null, expiresAt: null });
      mockAxiosHooks.mockReturnValue([{
        data: {
          referralId: null,
        },
      }]);
    });

    test('renders but does not include referralCode in share URLs', () => {
      const coursePath = '/courses/testing-referrals';
      const { container } = render(
        <SocialShare
          coursePath={coursePath}
        />,
      );

      expect(container).toMatchSnapshot();

      const expectedLinks = [
        {
          baseUrl: 'linkedin.com',
          assert: (link: Element | null) => {
            const href = link?.getAttribute('href');
            expect(href).not.toBeNull();

            const urlParams = new URLSearchParams(href!.split('?')[1]);
            const encodedTargetUrl = urlParams.get('url');
            expect(encodedTargetUrl).not.toBeNull();
            if (encodedTargetUrl) {
              const targetUrl = new URL(decodeURIComponent(encodedTargetUrl));
              expect(targetUrl.searchParams.get('r')).toBeNull();
              expect(targetUrl.searchParams.get('utm_source')).toBe('referral');
              expect(targetUrl.searchParams.get('utm_campaign')).toBe('linkedin');
            }
          },
        },
        {
          baseUrl: 'twitter.com',
          assert: (link: Element | null) => {
            const href = link?.getAttribute('href');
            expect(href).not.toBeNull();

            const urlParams = new URLSearchParams(href!.split('?')[1]);
            const encodedTargetUrl = urlParams.get('url');
            expect(encodedTargetUrl).not.toBeNull();
            if (encodedTargetUrl) {
              const targetUrl = new URL(decodeURIComponent(encodedTargetUrl));
              expect(targetUrl.searchParams.get('r')).toBeNull();
              expect(targetUrl.searchParams.get('utm_source')).toBe('referral');
              expect(targetUrl.searchParams.get('utm_campaign')).toBe('twitter');
            }
          },
        },
        {
          baseUrl: 'facebook.com',
          assert: (link: Element | null) => {
            const href = link?.getAttribute('href');
            expect(href).not.toBeNull();
          },
        },
      ];

      expectedLinks.forEach(({ baseUrl, assert }) => {
        const link = container.querySelector(`a[href*="${baseUrl}"]`);
        assert(link);
      });
    });
  });
});
