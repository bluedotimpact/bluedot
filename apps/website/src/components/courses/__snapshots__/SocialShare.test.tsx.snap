// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SocialShare > when user is logged in > renders default as expected 1`] = `
<div>
  <div
    class="social-share flex flex-row gap-4"
  >
    <a
      class="social-share__link size-6"
      href="https://www.linkedin.com/shareArticle?mini=true&url=http%3A%2F%2Flocalhost%3A3000%2Fcourses%2Ffuture-of-ai%3Futm_source%3Dreferral%26utm_campaign%3Dlinkedin%26r%3Dtest-referral-123&text=I've just completed a free, 2-hour course on the future of AI and its impacts on society. Here are my takeaways:"
      tabindex="0"
      target="_blank"
    >
      <svg
        class="social-share__link-icon size-6"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"
        />
      </svg>
    </a>
    <a
      class="social-share__link size-6"
      href="https://twitter.com/intent/tweet?url=http%3A%2F%2Flocalhost%3A3000%2Fcourses%2Ffuture-of-ai%3Futm_source%3Dreferral%26utm_campaign%3Dtwitter%26r%3Dtest-referral-123&text=I've just completed a free, 2-hour course on the future of AI and its impacts on society. Here are my takeaways:"
      tabindex="0"
      target="_blank"
    >
      <svg
        class="social-share__link-icon size-6"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"
        />
      </svg>
    </a>
    <a
      class="social-share__link size-6"
      href="https://www.facebook.com/sharer/sharer.php"
      tabindex="0"
      target="_blank"
    >
      <svg
        class="social-share__link-icon size-6"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 256C512 114.6 397.4 0 256 0S0 114.6 0 256C0 376 82.7 476.8 194.2 504.5V334.2H141.4V256h52.8V222.3c0-87.1 39.4-127.5 125-127.5c16.2 0 44.2 3.2 55.7 6.4V172c-6-.6-16.5-1-29.6-1c-42 0-58.2 15.9-58.2 57.2V256h83.6l-14.4 78.2H287V510.1C413.8 494.8 512 386.9 512 256h0z"
        />
      </svg>
    </a>
  </div>
</div>
`;

exports[`SocialShare > when user is logged in > renders without optional args 1`] = `"https://www.linkedin.com/shareArticle?mini=true&url=http%3A%2F%2Flocalhost%3A3000%2Ffuture-of-ai%3Futm_source%3Dreferral%26utm_campaign%3Dlinkedin%26r%3Dtest-referral-123"`;

exports[`SocialShare > when user is logged in > renders without optional args 2`] = `"https://twitter.com/intent/tweet?url=http%3A%2F%2Flocalhost%3A3000%2Ffuture-of-ai%3Futm_source%3Dreferral%26utm_campaign%3Dtwitter%26r%3Dtest-referral-123"`;

exports[`SocialShare > when user is logged in > renders without optional args 3`] = `"https://www.facebook.com/sharer/sharer.php"`;

exports[`SocialShare > when user is logged out > renders but does not include referralCode in share URLs 1`] = `
<div>
  <div
    class="social-share flex flex-row gap-4"
  >
    <a
      class="social-share__link size-6"
      href="https://www.linkedin.com/shareArticle?mini=true&url=http%3A%2F%2Flocalhost%3A3000%2Fcourses%2Ftesting-referrals%3Futm_source%3Dreferral%26utm_campaign%3Dlinkedin"
      tabindex="0"
      target="_blank"
    >
      <svg
        class="social-share__link-icon size-6"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"
        />
      </svg>
    </a>
    <a
      class="social-share__link size-6"
      href="https://twitter.com/intent/tweet?url=http%3A%2F%2Flocalhost%3A3000%2Fcourses%2Ftesting-referrals%3Futm_source%3Dreferral%26utm_campaign%3Dtwitter"
      tabindex="0"
      target="_blank"
    >
      <svg
        class="social-share__link-icon size-6"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"
        />
      </svg>
    </a>
    <a
      class="social-share__link size-6"
      href="https://www.facebook.com/sharer/sharer.php"
      tabindex="0"
      target="_blank"
    >
      <svg
        class="social-share__link-icon size-6"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 256C512 114.6 397.4 0 256 0S0 114.6 0 256C0 376 82.7 476.8 194.2 504.5V334.2H141.4V256h52.8V222.3c0-87.1 39.4-127.5 125-127.5c16.2 0 44.2 3.2 55.7 6.4V172c-6-.6-16.5-1-29.6-1c-42 0-58.2 15.9-58.2 57.2V256h83.6l-14.4 78.2H287V510.1C413.8 494.8 512 386.9 512 256h0z"
        />
      </svg>
    </a>
  </div>
</div>
`;
